src/ranged.cpp

```diff
@@ -929,33 +1022,51 @@ static int do_aim( player &p, const std::vector<Creature *> &t, int cur_target,
     return cur_target;
 }
 
+static double approx_hit_chance( double dispersion, double range, double missed_by ) {
+    // This is essentially the inverse of what Creature::projectile_attack() does.
+
+    double missed_by_tiles = missed_by * occupied_tile_fraction;
+
+    //          T = (D**2 * (1 - cos V)) ** 0.5   (from iso_tangent)
+    //      cos V = 1 - T**2 / (2*D**2)
+    // 1 - V**2/2 = 1 - T**2 / (2*D**2)           (small angle approximation)
+    //          V = T / D
+    double shot_dispersion = ( range < 0.1 ? M_PI : missed_by_tiles / range ) * 180 * 60 / M_PI;
+    double sigma = dispersion / dispersion_sigmas;
+
+    // erf(x / (sigma * sqrt(2))) is the probability that a measurement
+    // of a normally distributed variable with standard deviation sigma
+    // lies in the range [-x..x]
+    return std::erf( shot_dispersion / ( sigma * 1.4142 ) );
+}
+
 static int print_aim_bars( const player &p, WINDOW *w, int line_number, item *weapon,
                            Creature *target, double predicted_recoil ) {
     // This is absolute accuracy for the player.
     // TODO: push the calculations duplicated from Creature::deal_projectile_attack() and
     // Creature::projectile_attack() into shared methods.
     // Dodge is intentionally not accounted for.
 
-    // Confidence is chance of the actual shot being under the target threshold,
-    // This simplifies the calculation greatly, that's intentional.
-    const double aim_level = p.get_weapon_dispersion( *weapon ) + predicted_recoil + p.recoil_vehicle();
+    const double dispersion = p.get_weapon_dispersion( *weapon ) + predicted_recoil + p.recoil_vehicle();
     const double range = rl_dist( p.pos(), target->pos() );
-    const double missed_by = aim_level * 0.00021666666666666666 * range;
-    const double hit_rating = missed_by;
-    const double confidence = 1 / hit_rating;
+
+    const double p_headshot = approx_hit_chance( dispersion, range, accuracy_headshot );
```

Comment:
I think this needs all the verbosity it can get since it's evidently not an obvious bit of code.

If you really want to restyle it, I won't object to you doing that in a separate PR after you have reapplied this one.